Fixed triton_support_test on rocm.

zoranjovanovic-ns opened the pull request:
📝 Summary of Changes
Updated functions from triton/support.c and triton/support_test to execute
correctly on ROCm.

🎯 Justification
support_test was failing on ROCm.

🚀 Kind of Contribution
Please remove what does not apply: 🐛 Bug Fix, 🧪 Tests

📊 Benchmark (for Performance Improvements)
Please measure and include speedups for one of the public HLOs in
`compiler/xla/tools/benchmarks/hlo/`.

🧪 Unit Tests:
Used existing triton/support_test 

🧪 Execution Tests:
What execution tests were added? For example, a new optimization should be
tested with an end-to-end execution test triggering the optimization and
asserting correctness. Please provide test cases running with at most 2 GPUs.

zoranjovanovic-ns committed:
Fixed triton_support_test on rocm.

dimitar-asenov commented on xla/backends/gpu/codegen/triton/support.cc:
```diff
@@ -377,7 +394,8 @@ CodegenDecision AreTypesSupportedByAlgUnsetDot(
 
-  auto supported_float_types = {BF16, F16, F32, F64, F8E5M2, F8E4M3FN};
+  auto supported_float_types = {BF16, F16, F32, F64, F8E4M3FN, F8E5M2,
+                                F8E4M3FNUZ, F8E5M2FNUZ};
```
Shouldn't the support for the extra new types depend on `gpu_version`?

dimitar-asenov commented on xla/backends/gpu/codegen/triton/support_test.cc:
```diff
@@ -643,11 +658,21 @@ ENTRY triton_computation {
+          data_type == PrimitiveType::F8E4M3FN ||
+          data_type == PrimitiveType::F8E5M2FNUZ ||
+          data_type == PrimitiveType::F8E4M3FNUZ))) {
+      fail_mode = ExpectedFailMode::kFailOrCrash;
```
Why `FailOrCrash`? What happens exactly? We should avoid this mode if possible and prefer something mroe specific.

dimitar-asenov commented on xla/backends/gpu/codegen/triton/support_test.cc:
```diff
@@ -701,7 +726,20 @@ ENTRY triton_computation {
+         data_type == PrimitiveType::F8E4M3FNUZ);
+  }
+
+  RunSupportTest(std::move(ti), /*output_tile_sizes=*/{1, 32}, cc,
```
Why `FailOrCrash`? What happens exactly? We should avoid this mode if possible and prefer something mroe specific.

dimitar-asenov commented on xla/backends/gpu/codegen/triton/support_test.cc:
```diff
@@ -2416,6 +2476,15 @@ ENTRY triton_computation {
+        (absl::c_linear_search(std::vector{S64, S32, S16, BF16, F16, F32},
+                                data_type)  &&
+         algorithm == xla::PrecisionConfig::ALG_DOT_F64_F64_F64)) {
+      fail_mode = ExpectedFailMode::kFailOrCrash;
```
Why `FailOrCrash`? What happens exactly? We should avoid this mode if possible and prefer something mroe specific.

dimitar-asenov reviewed (changes requested):
Thanks, I left you some comments. For the ones asking about `FailOrCrash`, please either adjust the code, or add a brief comment to the code explaining why `FailOrCrash` is the right choice.

zoranjovanovic-ns committed:
Review comments.

penpornk commented:
[Linux x86 GPU CI failed](https://github.com/openxla/xla/actions/runs/20316907944/job/58428882612?pr=35363). Could you please help take a look?
```
[ RUN      ] BinaryElementwiseTestSuite/BinaryElementwiseTest.IsTritonSupportedBinaryElementwise/f8e4m3fnuz_divide_100
LLVM ERROR: Unsupported bitwidth
/b/f/w/bazel-out/k8-opt/bin/xla/backends/gpu/codegen/triton/support_test.runfiles/xla/build_tools/ci/parallel_gpu_execute: line 74:    28 Aborted                 (core dumped) "$TEST_BINARY" $@
```

zoranjovanovic-ns commented:
Cuda tests are failing because of changes in emitter_helpers.cc - adding support for F8E5M2FNUZ and F8E4M3FNUZ data types.
Before changes, CreateTritonModule always failed (resulting in test failure) and now it manages to create the module, but after that Triton pipeline crashes because those types are not supported.
Basically, number of tests that used to fail (in XLA) now crashes (in Triton passes) and that results in test failure.
Triton passes do not always crash with F8E5M2FNUZ and F8E4M3FNUZ - sometimes just fail.
This, most probably, means that it will be necessary to use ExpectedFailMode::kFailOrCrash for those tests (some tests fail and some tests crash).
Will try to find some additional time tomorrow to fix the tests for Cuda.

penpornk commented:
Thank you for the update!

zoranjovanovic-ns committed:
Fixed issues with F8E4M3FNUZ and F8E5M2FNUZ and Cuda in support_test

dimitar-asenov reviewed (changes requested):
Thanks for looking into the issue.

Please do not use `GTEST_SKIP` or any other way to skip these tests. They should not be skipped, but rather the behavior that actually happens should be tested (if it is a crash then check for a crash).

The reason is this:
- We want the Triton support tests to precisely reflect the current known state of the implemenetation.
- If you skip a certain test combination, then we get a hole in our tests. Imagine that at a later time, Cuda starts supporting these types. Now our tests won't detect this in any way and we are very likely to forget to update them.
- Instead, if you specifically test that these types do not work under the CUDA (e.g. by testing for a crash) and at some later point Cuda implements them, then our tests will detect the discrepancy and we will update them. This will ensure that we're testing the full possible extend of the Triton code-gen capabilities.

zoranjovanovic-ns commented:
@dimitar-asenov As far as I can see, the problem is that some operations just happen to work on Cuda Triton for F8E4M3FNUZ and F8E5M2FNUZ data types.

It seems to me that, with current tests logic, only way to have some tests to pass is to modify some of the functions from support.c (e.g. IsTritonSupportedConversion) to return **true** for those data types. Otherwise, test expects those cases to fail (it does not matter with fail or crash) but some pass - resulting in test failures.

In my opinion, we should not modify IsTritonSupportedConversion and similar functions to return true for cuda and those data types, because there is no intentional support for F8E4M3FNUZ and F8E5M2FNUZ in cuda Triton.

dimitar-asenov commented:
Thanks for the details.

I see two options:
- If the generated code really works and that's not a bug, then we should change `IsTritonSupportedConversion`.
- If the generated code works, but that's more of an accident and seems broken, we can change the implementation of the codegen to check for these types under CUDA and if encountered raise and explicit error. This will enable us to properly write the test without modifying `IsTritonSupportedConversion`.

Could you please clarify which exact tests seem to pass? I'm not that familiar with the FP8 support on CUDA and Triton, what do you mean when you say that it's unintentional?

zoranjovanovic-ns commented:
> I'm not that familiar with the FP8 support on CUDA and Triton, what do you mean when you say that it's unintentional?

I meant that generated code works by accident, or because conversion from one fp8 type to the same fp8 does not require much code :) .

Will check second option from your comment.

zoranjovanovic-ns committed:
Review comments.

zoranjovanovic-ns committed:
Fix rocm reduce subtests.

zoranjovanovic-ns commented:
Hi @dimitar-asenov , updated PR.

zoranjovanovic-ns committed:
Removed skip for cuda and fnuz data types in support_test.

zoranjovanovic-ns committed:
Used IsTritonSupportedInstruction to verify that instructions are
actually supported by Triton.
Modified support_test to follow new implementation.

zoranjovanovic-ns committed:
Modify support_test for Rocm to reflect new implementation.

dimitar-asenov commented:
The commit you linked is independent to the cleanups we wanted to do. Regarding those we explored one path internally but that turned out to not be quite what we want, so now we have to do something else. Sorry for the delay.

zoranjovanovic-ns commented:
> The commit you linked is independent to the cleanups we wanted to do. Regarding those we explored one path internally but that turned out to not be quite what we want, so now we have to do something else. Sorry for the delay.

Thank you for the fast reply.
Any estimate on when this refactoring will be ready (to be able to plan my activities on upstreaming changes)?

dimitar-asenov commented:
@zoranjovanovic-ns The refactoring is in:
https://github.com/openxla/xla/pull/37640

Please rebase on top and adjust this PR. To reiterate the important points:
- The code-gen layer should not call into any `IsSupported...` function.
- Your code-gen should avoid explicit checks of support conditions and just lets lower levels  (Triton/LLVM) fail gracefully.
- There are two cases where writing extra checks in your code-gen is reasonable:
  - If the check is necessary to decide how to codegen.
  - If the lower levels fail with a crash.
  - If you write checks here, they should not call into the general `IsSupport...` function.
- The support test should accurately reflect what is supported, what fails gracefully, and what crashes.

Hope this helps.
